### PR TITLE
Fix dynamic attriubtes

### PR DESCRIPTION
**Pull request by KrystianKedron**
#99

### CODE REVIEW

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonDeviceHL.xtend [hunk 1]
```diff
@@ -1,473 +1,474 @@
...
+«IF !attr.readExcludedStates.empty || !attr.writeExcludedStates.empty»    «attributeMethodStateMachineHL(cls, attr)»«ENDIF»
+«ENDIF»«ENDFOR»
+    '''
+    //====================================================
+    //    Dynamic Attributes
+    //====================================================
+    def pythonDynamicAttributes(PogoDeviceClass cls)  '''
+«FOR attr: cls.dynamicAttributes»«IF isTrue(attr.status.concreteHere)»
+«IF attr.isRead»    «readAttributeMethodHL(cls, attr, true)»«ENDIF»
+«IF attr.isWrite»    «writeAttributeMethodHL(cls, attr, true)»«ENDIF»
```
Comment (KrystianKedron) [inline fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonDeviceHL.xtend:338]: change

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonDeviceHL.xtend [hunk 2]
```diff
@@ -1,473 +1,474 @@
...
+«ENDIF»
+«ENDFOR»«ENDIF»
+    '''
+    //====================================================
+    //    Attributes
+    //====================================================
+    def pythonAttributes(PogoDeviceClass cls)  '''
+«FOR attr: cls.attributes»«IF isTrue(attr.status.concreteHere)»
+«IF attr.isRead»    «readAttributeMethodHL(cls, attr, false)»«ENDIF»
+«IF attr.isWrite»    «writeAttributeMethodHL(cls, attr, false)»«ENDIF»
```
Comment (KrystianKedron) [inline fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonDeviceHL.xtend:328]: change

**Comment by KrystianKedron** [comment]
Example generated code:

```
    def initialize_dynamic_attributes(self):
        self.debug_stream("In initialize_dynamic_attributes()")
        # PROTECTED REGION ID(Test.initialize_dynamic_attributes) ENABLED START #
        # PROTECTED REGION END #    //  Test.initialize_dynamic_attributes

        """   Example to add dynamic attributes
           Copy inside the following code to protected area to instantiate at startup."""
        """    For Attribute testDynamicScalar
        mytestDynamicScalar = PyTango.Attr('MytestDynamicScalar', PyTango.DevDouble, PyTango.READ_WRITE)
        self.add_attribute(mytestDynamicScalar,Test.read_testDynamicScalar, Test.write_testDynamicScalar, None)
        self.attr_testDynamicScalar_read = 0.0
        """

    def read_testDynamicScalar(self, attr):
        # PROTECTED REGION ID(Test.testDynamicScalar_read) ENABLED START #
        pass
        # PROTECTED REGION END #    //  Test.testDynamicScalar_read
        """   Example how to read dynamic attribute
        Copy inside the following code to protected area
        attr.set_value(self.attr_testDynamicScalar_read)
        """

    def write_testDynamicScalar(self, w_attr):
        # PROTECTED REGION ID(Test.testDynamicScalar_write) ENABLED START #
        """Set the testDynamicScalar attribute."""
        pass
        # PROTECTED REGION END #    //  Test.testDynamicScalar_write
        """   Example how to write dynamic attribute
        Copy inside the following codr_testDynamicScalar_read = w_attr.get_write_value()
        """e to protected area
        name = w_attr.get_name()
        self.att
```

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonDeviceHL.xtend [hunk 3]
```diff
@@ -1,473 +1,475 @@
-//+======================================================================
```
Comment (ajoubertza) [inline fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonDeviceHL.xtend:1]: Did the end-of-line characters change on this file?  I suggest that is fixed, so that the diff doesn't show unnecessary changes.  It is hard to tell which lines really changed.

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend [hunk 4]
```diff
@@ -415,6 +415,13 @@ class PythonUtils {
                 «ELSE»
                 pass
                 «ENDIF»
+                «IF isDynamic»
+                """   Example how to write dynamic attribute
+                Copy inside the following code to protected area
+                name = w_attr.get_name()
+                self.attr_«attribute.name»_read = w_attr.get_write_value()
```
Comment (ajoubertza) [inline fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend:422]: The example attribute name ending with `_read` is a bit confusing for a read/write attribute.  I see that is what the old dynamic example code proposes in `initialize_dynamic_attributes`.  Consider dropping the `_read` suffix from the read, write and initialise examples.

**Review by ajoubertza — COMMENTED** [review]
Thanks @KrystianKedron.  The PyTango implementation of the generated code looks OK.  I have a few other comments related to readability.  Also consider what the code will look like after someone with a real implementation has regenerated.

**Comment by ajoubertza** [comment]
> Example of generated code:
> 
> ```
>     def initialize_dynamic_attributes(self):
>         self.debug_stream("In initialize_dynamic_attributes()")
>         # PROTECTED REGION ID(Test.initialize_dynamic_attributes) ENABLED START #
>         # PROTECTED REGION END #    //  Test.initialize_dynamic_attributes
> 
>         """   Example to add dynamic attributes
>            Copy inside the following code to protected area to instantiate at startup."""
>         """    For Attribute testDynamicScalar
>         mytestDynamicScalar = PyTango.Attr('MytestDynamicScalar', PyTango.DevDouble, PyTango.READ_WRITE)
>         self.add_attribute(mytestDynamicScalar,Test.read_testDynamicScalar, Test.write_testDynamicScalar, None)
>         self.attr_testDynamicScalar_read = 0.0
>         """
> 
>     def read_testDynamicScalar(self, attr):
>         # PROTECTED REGION ID(Test.testDynamicScalar_read) ENABLED START #
>         pass
>         # PROTECTED REGION END #    //  Test.testDynamicScalar_read
>         """   Example how to read dynamic attribute
>         Copy inside the following code to protected area
>         attr.set_value(self.attr_testDynamicScalar_read)
>         """
> 
>     def write_testDynamicScalar(self, w_attr):
>         # PROTECTED REGION ID(Test.testDynamicScalar_write) ENABLED START #
>         """Set the testDynamicScalar attribute."""
>         pass
>         # PROTECTED REGION END #    //  Test.testDynamicScalar_write
>         """   Example how to write dynamic attribute
>         Copy inside the following codr_testDynamicScalar_read = w_attr.get_write_value()
>         """e to protected area
>         name = w_attr.get_name()
>         self.att
> ```

In the last method, `write_testDynamicScalar`, the output looks wrong after.  Syntax error after the last `"""`.  I think it is probably an error in the comment, and not the code, but please check and update.

P.S. you can get the markdown to have Python syntax colouring like this:
image

**Comment by KrystianKedron** [comment]
1.  
> Also consider what the code will look like after someone with a real implementation has regenerated.

Yes, I thought about that, my code only added the w_attr to the write_testDynamicScalar and this is the only change that can break some already generated code. But this is in the scope of this task, and for now, I don't see a different possibility to solve this issue. 
What do you think? Do you see different risks?

2.  
> > Example of generated code:
> > ```
> >     def initialize_dynamic_attributes(self):
> >         self.debug_stream("In initialize_dynamic_attributes()")
> >         # PROTECTED REGION ID(Test.initialize_dynamic_attributes) ENABLED START #
> >         # PROTECTED REGION END #    //  Test.initialize_dynamic_attributes
> > 
> >         """   Example to add dynamic attributes
> >            Copy inside the following code to protected area to instantiate at startup."""
> >         """    For Attribute testDynamicScalar
> >         mytestDynamicScalar = PyTango.Attr('MytestDynamicScalar', PyTango.DevDouble, PyTango.READ_WRITE)
> >         self.add_attribute(mytestDynamicScalar,Test.read_testDynamicScalar, Test.write_testDynamicScalar, None)
> >         self.attr_testDynamicScalar_read = 0.0
> >         """
> > 
> >     def read_testDynamicScalar(self, attr):
> >         # PROTECTED REGION ID(Test.testDynamicScalar_read) ENABLED START #
> >         pass
> >         # PROTECTED REGION END #    //  Test.testDynamicScalar_read
> >         """   Example how to read dynamic attribute
> >         Copy inside the following code to protected area
> >         attr.set_value(self.attr_testDynamicScalar_read)
> >         """
> > 
> >     def write_testDynamicScalar(self, w_attr):
> >         # PROTECTED REGION ID(Test.testDynamicScalar_write) ENABLED START #
> >         """Set the testDynamicScalar attribute."""
> >         pass
> >         # PROTECTED REGION END #    //  Test.testDynamicScalar_write
> >         """   Example how to write dynamic attribute
> >         Copy inside the following codr_testDynamicScalar_read = w_attr.get_write_value()
> >         """e to protected area
> >         name = w_attr.get_name()
> >         self.att
> > ```
> 
> In the last method, `write_testDynamicScalar`, the output looks wrong after. Syntax error after the last `"""`. I think it is probably an error in the comment, and not the code, but please check and update.
> 
> P.S. you can get the markdown to have Python syntax colouring like this:
> image

Yes, that was a typo. Good catch, the proper example below (thx for markdown):
```python
    def initialize_dynamic_attributes(self):
        self.debug_stream("In initialize_dynamic_attributes()")
        # PROTECTED REGION ID(Test.initialize_dynamic_attributes) ENABLED START #
        # PROTECTED REGION END #    //  Test.initialize_dynamic_attributes

        """   Example to add dynamic attributes
           Copy inside the following code to protected area to instantiate at startup."""
        """    For Attribute testDynamicScalar
        mytestDynamicScalar = PyTango.Attr('MytestDynamicScalar', PyTango.DevDouble, PyTango.READ_WRITE)
        self.add_attribute(mytestDynamicScalar,Test.read_testDynamicScalar, Test.write_testDynamicScalar, None)
        self.attr_testDynamicScalar_read = 0.0
        """

    def read_testDynamicScalar(self, attr):
        # PROTECTED REGION ID(Test.testDynamicScalar_read) ENABLED START #
        pass
        # PROTECTED REGION END #    //  Test.testDynamicScalar_read
        """   Example how to read dynamic attribute
        Copy inside the following code to protected area
        attr.set_value(self.attr_testDynamicScalar_read)
        """

    def write_testDynamicScalar(self, w_attr):
        # PROTECTED REGION ID(Test.testDynamicScalar_write) ENABLED START #
        """Set the testDynamicScalar attribute."""
        pass
        # PROTECTED REGION END #    //  Test.testDynamicScalar_write
        """   Example how to write dynamic attribute
        Copy inside the following code to protected area
        name = w_attr.get_name()
        self.attr_testDynamicScalar_read = w_attr.get_write_value()
        """
```

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend (hunk 4 above)
Comment (KrystianKedron) [reply fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend:422]: I like your proposition, so this will be a good candidate for the next PR. 
My plan is to create a new task to discuss what this suffix should be changed (I don't think this is in the scope of this task).
But if you want we can do this in this current PR. If not, let's move the discussion to the next task that I will create, 

What do you think?

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonDeviceHL.xtend (hunk 3 above)
Comment (KrystianKedron) [reply fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonDeviceHL.xtend]: I using the IntelliJ and this IDE has a problem with committing those .xtend files. 
But I agree this formatting change isn't something that should be added to the PR.
I used a different way of creating the commit and now the changes should be better visible.

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend (hunk 4 above)
Comment (ajoubertza) [reply fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend:422]: Yes, it can wait for another PR.

**Comment by ajoubertza** [comment]
> Yes, I thought about that, my code only added the w_attr to the write_testDynamicScalar and this is the only change that can break some already generated code. But this is in the scope of this task, and for now, I don't see a different possibility to solve this issue.
> What do you think? Do you see different risks?

I don't think your PR will cause things to break.  My issue is that the code will look a little messy, once there is a real implementation.

Initial code:
```python

    def write_testDynamicScalar(self, w_attr):
        # PROTECTED REGION ID(Test.testDynamicScalar_write) ENABLED START #
        """Set the testDynamicScalar attribute."""
        pass
        # PROTECTED REGION END #    //  Test.testDynamicScalar_write
        """   Example how to write dynamic attribute
        Copy inside the following code to protected area
        name = w_attr.get_name()
        self.attr_testDynamicScalar_read = w_attr.get_write_value()
        """
```

Add custom implementation (I would probably delete the example code):
```python

    def write_testDynamicScalar(self, w_attr):
        # PROTECTED REGION ID(Test.testDynamicScalar_write) ENABLED START #
        """Set the testDynamicScalar attribute."""
        self.attr_testDynamicScalar_read = w_attr.get_write_value()
        # PROTECTED REGION END #    //  Test.testDynamicScalar_write
```

Re-generated code (preserving protected regions):
```python

    def write_testDynamicScalar(self, w_attr):
        # PROTECTED REGION ID(Test.testDynamicScalar_write) ENABLED START #
        """Set the testDynamicScalar attribute."""
        self.attr_testDynamicScalar_read = w_attr.get_write_value()
        # PROTECTED REGION END #    //  Test.testDynamicScalar_write
        """   Example how to write dynamic attribute
        Copy inside the following code to protected area
        name = w_attr.get_name()
        self.attr_testDynamicScalar_read = w_attr.get_write_value()
        """
```


------------------------

Alternative, if the example code is put inside the protected region when generated:

Initial code:
```python

    def write_testDynamicScalar(self, w_attr):
        # PROTECTED REGION ID(Test.testDynamicScalar_write) ENABLED START #
        """Set the testDynamicScalar attribute."""
        pass
        """   Example how to write dynamic attribute
        Copy inside the following code to protected area
        name = w_attr.get_name()
        self.attr_testDynamicScalar_read = w_attr.get_write_value()
        """
        # PROTECTED REGION END #    //  Test.testDynamicScalar_write
```

Add custom implementation (I would probably delete the example code):
```python

    def write_testDynamicScalar(self, w_attr):
        # PROTECTED REGION ID(Test.testDynamicScalar_write) ENABLED START #
        """Set the testDynamicScalar attribute."""
        self.attr_testDynamicScalar_read = w_attr.get_write_value()
        # PROTECTED REGION END #    //  Test.testDynamicScalar_write
```

Re-generated code (preserving protected regions):
```python

    def write_testDynamicScalar(self, w_attr):
        # PROTECTED REGION ID(Test.testDynamicScalar_write) ENABLED START #
        """Set the testDynamicScalar attribute."""
        self.attr_testDynamicScalar_read = w_attr.get_write_value()
        # PROTECTED REGION END #    //  Test.testDynamicScalar_write
```

**Comment by KrystianKedron** [comment]
> I don't think your PR will cause things to break. My issue is that the code will look a little messy, once there is a real implementation.

Sure, agree. Simply speaking I based on the current solution in the initialize_dynamic_attributes method.

But I think a better solution is to have the generated code in the way like you mention (in PR region):

```python
    def write_dynamicSpectrumAttribute(self, w_attr):
        # PROTECTED REGION ID(Test.dynamicSpectrumAttribute_write) ENABLED START #
        """Set the dynamicSpectrumAttribute attribute."""
        name = w_attr.get_name()
        self.attr_dynamicSpectrumAttribute_read = w_attr.get_write_value()
        # PROTECTED REGION END #    //  Test.dynamicSpectrumAttribute_write

    def read_dynamicImageAttribute(self, attr):
        # PROTECTED REGION ID(Test.dynamicImageAttribute_read) ENABLED START #
        """Return the dynamicImageAttribute_read attribute."""
        attr.set_value(self.attr_dynamicImageAttribute_read)
        return attr
        # PROTECTED REGION END #    //  Test.dynamicImageAttribute_read
```

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend [hunk 5]
```diff
@@ -424,13 +429,18 @@ class PythonUtils {
...
+    def readAttributeMethodHL(PogoDeviceClass cls, Attribute attribute, boolean isDynamic) '''
+        def read_«attribute.name»(self«IF isDynamic», attr«ENDIF»):
                 «IF cls.description.filestogenerate.toLowerCase.contains("protected regions")»
                 «openProtectedAreaHL(cls, attribute.name + "_read")»
+                «IF isDynamic==false»
                 """Return the «attribute.name» attribute."""
                 return self._«attribute.pythonAttributeVariableNameHL»
+                «ELSE»
+                """Return the «attribute.name»_read attribute."""
+                attr.set_value(self.attr_«attribute.name»_read)
```
Comment (ajoubertza) [inline fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend:441]: If we are going to include a real implementation that is active on first generation, then the variable we are using, `self.attr_«attribute.name»_read`, must exist.   I see that the initialisation of the variable, created by by [`dynamicAttributeDefaultValueExample`](https://github.com/tango-controls/pogo/blob/410f63460b632a86963bacc738d523fe2591d1d6/fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonDeviceHL.xtend#L358), will be inside a docstring, so the variable won't exist.  

Either put the implementations for read and write inside docstrings, or take the example initialisation out of the docstrings.  I think keeping everything in docstrings would be more consistent with the previous behaviour of Pogo.

**Review by ajoubertza — COMMENTED** [review]
(Sorry for the delayed reply)

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend (hunk 5 above)
Comment (KrystianKedron) [reply fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend:441]: Thx,
Ok so to summarize our review:
- in the next task I will apply the changes to remove the _read suffix
- also the current approach to providing the hint on how to implement the dynamic attributes isn't perfect and has a problem as you describe in your comment. Solving this in the pogo implementation will be required an additional method to check the code inside of the comments (```code```) which should be generated only once. And this isn't trivial. What is your proposition about this problem?

Also getting back to current PR, what are your opinion about the current state of this PR, when the code will be ready to be merged in your opinion, and what I should do in this PR? Because I meet the requirements and provide the solution to the problem in the ticket. I follow the current way that we have in the pogo.

Currently what left to do is to get back to the previous solution to add the suggestion about the implementation of the dynamic attributes in the comment block (if this is accepted way by you).

Comment (ajoubertza) [reply fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend:441]: > Solving this in the pogo implementation will be required an additional method to check the code inside of the comments (code) which should be generated only once. And this isn't trivial. What is your proposition about this problem?

The problem is that the initialisation code is a commented-out (docstring) example, while the read/write implementations are not commented out and use an undefined variable.  My proposition is not to do any extra checking - you are right that it wouldn't be trivial.   Option 1:  generate code that isn't commented out and works in both the initialiser and the read/write implementations.  Option 2:  generate code that is commented out in both the initialiser and the read/write implementations.

> Also getting back to current PR, what are your opinion about the current state of this PR, when the code will be ready to be merged in your opinion, and what I should do in this PR? Because I meet the requirements and provide the solution to the problem in the ticket. I follow the current way that we have in the pogo.

I think it has an issue that need to be fixed before it can be merged.  It doesn't meet the requirements because (I think) the generated code will raise an exception if a client tries to read a dynamic attribute before writing to it.  Have you tested the generated code, and tried to read an attribute before writing it?

> Currently what left to do is to get back to the previous solution to add the suggestion about the implementation of the dynamic attributes in the comment block (if this is accepted way by you).

In the current PR, I think the easiest so solution is to leave the code where it is, but put it inside a docstring, so that it will not be executed (and raise an exception).  See my suggestions in the additional comments.

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend [hunk 6]
```diff
@@ -405,12 +405,17 @@ class PythonUtils {
...
+    def writeAttributeMethodHL(PogoDeviceClass cls, Attribute attribute, boolean isDynamic) '''
+        def write_«attribute.name»(self, «IF isDynamic»w_attr«ELSE»value«ENDIF»):
                 «IF cls.description.filestogenerate.toLowerCase.contains("protected regions")»
                 «openProtectedAreaHL(cls, attribute.name + "_write")»
                 """Set the «attribute.name» attribute."""
+                «IF isDynamic==false»
                 pass
+                «ELSE»
+                name = w_attr.get_name()
+                self.attr_«attribute.name»_read = w_attr.get_write_value()
```
Comment (ajoubertza) [inline fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend:417]: Since the `self.attr_«attribute.name»_read` is not initialised in the standard code generated by Pogo, suggest we comment this code out.  It would work, if uncommented, but since the read attribute method might not work, it seems better to have both read and write method commented out for consistency.
```suggestion
                """Example implementation:
                name = w_attr.get_name()
                self.attr_«attribute.name»_read = w_attr.get_write_value()
                """
```

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend [hunk 7]
```diff
@@ -424,13 +429,18 @@ class PythonUtils {
...
+        def read_«attribute.name»(self«IF isDynamic», attr«ENDIF»):
                 «IF cls.description.filestogenerate.toLowerCase.contains("protected regions")»
                 «openProtectedAreaHL(cls, attribute.name + "_read")»
+                «IF isDynamic==false»
                 """Return the «attribute.name» attribute."""
                 return self._«attribute.pythonAttributeVariableNameHL»
+                «ELSE»
+                """Return the «attribute.name»_read attribute."""
+                attr.set_value(self.attr_«attribute.name»_read)
+                return attr
```
Comment (ajoubertza) [inline fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend:442]: Since `self.attr_«attribute.name»_read` might not exist, suggest we comment out its use, and indicate it is an example.  We still need to return something - one option is the default value for the type of attribute, similar to line 446.
```suggestion
                """Example implementation:
                attr.set_value(self.attr_«attribute.name»_read)
                return attr
                """
                return «attribute.defaultValueDim»
```


This option might work too, although I'm not sure what value we will get from `attr` if we don't set it:
```suggestion
                """Example implementation:
                attr.set_value(self.attr_«attribute.name»_read)
                """
                return attr
```

**Review by ajoubertza — COMMENTED** [review]
I've made some suggestions.  If you use them, please test - I have not tried the code.

File: fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend (hunk 7 above)
Comment (KrystianKedron) [reply fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend:442]: Thx for your answer.

Exactly, that I mean by:
> Currently what left to do is to get back to the previous solution to add the suggestion about the implementation of the dynamic attributes in the comment block (if this is accepted way by you).

And this is compliant with:
> Option 2: generate code that is commented out in both the initializer and the read/write implementations.

I will follow that way, and test generated device server,

Comment (KrystianKedron) [reply fr.esrf.tango.generator.xtend/src/fr/esrf/tango/pogo/generator/python/PythonUtils.xtend]: Hi,

After my tests I update this PR with working solution:

```python
    def read_dynamicImageAttribute(self, attr):
        # PROTECTED REGION ID(Test.dynamicImageAttribute_read) ENABLED START #
        """Return the dynamicImageAttribute_read attribute."""
        """Example implementation:
        Hint: self.attr_dynamicImageAttribute should be define on the top level
        (initialize_dynamic_attributes, init_device method or constructor)
        attr.set_value(self.attr_dynamicImageAttribute)
        return attr
        """
        attr.set_value([[0.0]])
        return attr
        # PROTECTED REGION END #    //  Test.dynamicImageAttribute_read

    def write_dynamicImageAttribute(self, w_attr):
        # PROTECTED REGION ID(Test.dynamicImageAttribute_write) ENABLED START #
        """Set the dynamicImageAttribute attribute."""
        """Example implementation:
        name = w_attr.get_name()
        Hint: self.attr_dynamicImageAttribute should be define on the top level
        (initialize_dynamic_attributes, init_device method or constructor)
        self.attr_dynamicImageAttribute = w_attr.get_write_value()
        """
        pass
        # PROTECTED REGION END #    //  Test.dynamicImageAttribute_write
```

attr.set_value([[0.0]]) this is a attr.set_value(«attribute.defaultValueDim»)

I'm open to your feedback and any information on what I can also improve.

**Review by ajoubertza — APPROVED** [review]
Looks good!
